kubetest/gcedind/gce_dind.go

```diff
@@ -0,0 +1,622 @@
+			return err
+		}
+	}
+	return nil
+}
+
+// saveWorkerNodeLogs collects docker state, service logs, and Kubernetes
+// system pod logs from all worker node containers on the GCE instance,
+// and saves the logs in a local artifacts directory.
+func (d *Deployer) saveWorkerNodeLogs(artifactsDir string) error {
+	nodes, err := d.detectNodeContainers(kubeNodePrefix)
+	if err != nil {
+		return err
+	}
+	for _, node := range nodes {
+		if err := d.saveNodeLogs(node, artifactsDir, systemdServices, nodeKubeLogs); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// detectNodeContainers creates a list of names for either all master or all
+// worker node containers. It does this by running 'docker ps' and searching
+// for container names that begin with a specified name prefix.
+func (d *Deployer) detectNodeContainers(namePrefix string) ([]string, error) {
+	log.Printf("Looking for container names beginning with '%s'", namePrefix)
+	o, err := d.getOutput("docker ps")
+	if err != nil {
+		return nil, err
+	}
+	var nodes []string
+	trimmed := strings.TrimSpace(string(o))
+	if trimmed != "" {
+		lines := strings.Split(trimmed, "\n")
+		for _, line := range lines {
+			if strings.Contains(line, "ago") {
+				fields := strings.Fields(line)
+				name := fields[len(fields)-1]
+				if strings.Contains(name, namePrefix) {
+					nodes = append(nodes, name)
+				}
+			}
+		}
+	}
+	return nodes, nil
+}
+
+// saveNodeLogs collects docker state, service logs, and Kubernetes
+// system pod logs for a given node container, and saves the logs in a local
+// artifacts directory.
+func (d *Deployer) saveNodeLogs(node string, artifactsDir string, services []string, logFiles []string) error {
+	log.Printf("Saving logs from node container %s", node)
+
+	// Create directory for node container artifacts
+	logDir := filepath.Join(artifactsDir, node)
+	if err := d.run("mkdir -p " + logDir); err != nil {
+		return err
+	}
+
+	// Save docker state for this node
+	nodeExecCmder := d.newNodeExecCmder(node)
+	for _, dockerCommand := range dockerCommands {
+		if err := d.execCmdSaveLog(nodeExecCmder, dockerCommand.cmd, logDir, dockerCommand.logFile); err != nil {
+			return err
+		}
+	}
+
+	// Copy service logs from the node container
+	for _, svc := range services {
+		cmd := fmt.Sprintf("journalctl --output=cat -u %s.service", svc)
+		logFile := fmt.Sprintf("%s.log", svc)
+		if err := d.execCmdSaveLog(nodeExecCmder, cmd, logDir, logFile); err != nil {
+			return err
+		}
+	}
+
+	// Copy log files for kube system pods that are running in this node
+	// container. The 'docker cp ...' doesn't support wild carding, so this
+	// is accomplished by first running 'find' on the node container to
+	// find matching log files, and then 'docker cp ...' is run to copy
+	// each individual log file from the container node.
+	for _, logFile := range logFiles {
+		log.Printf("Saving %s", logFile)
+		cmd := fmt.Sprintf("find %s -name %s -print", nodeLogDir, logFile)
```

Comment:
@ixdy how much do we care about injections in kubetest? Eventually these logFiles might become parameterized.